README
NAME
    Csistck - Perl system consistency check framework

SYNOPSIS
        use Csistck;
    
        for my $host (qw/a b/) {
            host $host => role('test');
        }

        host 'c' => role('test');
    
        role 'test' => 
            template(".files/test.tt", "/tmp/test", { text => "Some text here" }),
            permission("/tmp/test*", mode => '0777', uid => 100, gid => 100);
    
        check;

    The script can then be called directly, using command line arguements
    below

DESCRIPTION
    Csistck is a small Perl framework for writing scripts to maintain system
    configuration and consistency. The focus of csistck is to stay
    lightweight, simple, and flexible.

METHODS
  option($name, $value)
    Set option to specified value.

   Available Options
    *  pkg_type [string]

       Set package type

    *  domain_name [string]

       Set default domain name to append to hosts

  host($hostname, [@tests]);
    Append test or array of tests to host definition.

        host 'hostname' => noop(1), noop(1);
        host 'hostname' => noop(0);

    Returns a reference to the host object.

  role($rolename, [@tests]);
    Append test or array of tests to role definition.

        role 'test' => noop(0);
        host 'hostname' => role('test');

    Returns a reference to the role object.

  noop($return)
    "No operation" test, used only for testing or placeholders.

        role 'test' => noop(1);

  file($glob, $target)
    Copy files matching file glob pattern to target directory.

        role 'test' => file("lighttpd/app/*.conf", "/etc/lighttpd");

    See Csistck::Test::File

  template($template, $target, [%args])
    Process file $template as a Template Toolkit template, output to path
    $target. Hashref %args is passed to the template processor.

        role 'test' => template("sys/motd", "/etc/motd", { hot => 'dog' });

    See Csistck::Test::Template

  permission($glob, %args)
    Change permissions on files matching file glob pattern

        role 'test' => permission("/etc/couchdb/*", {
          mode => '0640',
          uid => 130,
          gid => 130
        });

    See Csistck::Test::Permission

  script($script, [@arguments])
    Call script with specified arguments

        role 'test' => script("apache2/mod-check", "rewrite");

    See Csistck::Test::Script

  pkg($package, [$type])
    Check for package using system package manager.

        option 'pkg_type' => 'dpkg';
        role 'test' => pkg("lighttpd");

    See Csistck::Test::Pkg

SCRIPT USAGE
    Scripts based on csistck will run in an interactive mode by default. The
    following command line options are recognized in a csistck based script

    *   --[no]repair

        Automate repair mode, do not run in interactive mode.

    *   --[no]verbose

        Toggle verbose reporting of events

    *   --[no]debug

        Toggle debug reporting of events

    *   --[no]quiet

        Toggle event reporting of errors

AUTHOR
    Anthony Johnson, <anthony@ohess.org>

COPYRIGHT AND LICENSE
    Copyright (c) 2011 Anthony Johnson

    Permission is hereby granted, free of charge, to any person obtaining a
    copy of this software and associated documentation files (the
    "Software"), to deal in the Software without restriction, including
    without limitation the rights to use, copy, modify, merge, publish,
    distribute, sublicense, and/or sell copies of the Software, and to
    permit persons to whom the Software is furnished to do so, subject to
    the following conditions:

    The above copyright notice and this permission notice shall be included
    in all copies or substantial portions of the Software.

    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
    OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
    MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
    IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
    CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
    TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
    SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.



Changes
Revision history for Csistck.
 
 0.08 3-14-2012
     - Added base class for extending configurable roles
     - More pkg test managers
     - Basic colored output
     - Fixed pkg and permission tests
     - Updated POD and added CLI options
 0.01  Thu Jan 27 14:57:13 2011
	- original version; created by h2xs 1.23 with options
		-AXn Csistck



lib/Csistck/Test.pm
package Csistck::Test;

use 5.010;
use strict;
use warnings;

use base 'Exporter';
use Csistck::Oper;

sub new {
    my $class = shift;
    my %args = @_;
    my $args = \%args;
    
    # Build Test object to bless and return
    my $self = {};
    $self->{CHECK} = sub { };
    $self->{REPAIR} = sub { };
    $self->{DIFF} = undef;
    $self->{DESC} = "Unidentified test";
    bless $self, $class;
    return $self;
}

# Check, use eval to catch fatal errors in test
sub check {
    my $self = shift;

    die ("Not a code reference")
      unless (ref $self->{CHECK} eq "CODE");
    
    # Execute code reference in eval, return response
    Csistck::Oper::info("$self->{DESC}");
    eval { &{$self->{CHECK}}; };

    if ($@) {
        my $error = $@;
        $error =~ s/ at [A-Za-z0-9\/\_\-\.]+ line [0-9]+.\n//;
        Csistck::Oper::error("$self->{DESC}: $error");
        return 0;
    }
    else {
        return 1;
    }   
}

# Repair, use eval to catch fatal errors in repair
sub repair {
    my $self = shift;

    die ("Not a code reference")
      unless ($self->has_repair());
    
    # Execute code reference in eval, return response
    Csistck::Oper::info("Repairing $self->{DESC}");
    eval { &{$self->{REPAIR}}; };

    if ($@) {
        my $error = $@;
        $error =~ s/ at [A-Za-z0-9\/\_\-\.]+ line [0-9]+.\n//;
        Csistck::Oper::error("Repairing $self->{DESC}: $error");
        return 0;
    }
    else {
        return 1;
    }   
}

# Returns if test object has proper repair action
sub has_repair {
    my $self = shift;

    return 1
      if (ref $self->{REPAIR} eq 'CODE');

    return 0;
}

# Diff, show some form of diff for interactive mode
sub diff {
    my $self = shift;

    # No diff code defined
    unless (defined $self->{DIFF}) {
        return 1;
    }

    die ("Not a code reference")
      unless ($self->has_diff());

    # Execute code reference in eval, return response
    eval { &{$self->{DIFF}}; };

    return 1;
}

# Returns if test object has proper diff action
sub has_diff {
    my $self = shift;

    return 1
      if (ref $self->{DIFF} eq 'CODE');

    return 0;
}

1;


lib/Csistck/Test/Pkg.pm
package Csistck::Test::Pkg;

use 5.010;
use strict;
use warnings;

use base 'Csistck::Test';
use Csistck::Oper qw/debug/;
use Csistck::Config qw/option/;

our @EXPORT_OK = qw/pkg/;

use Digest::MD5;
use File::Basename;

# Conditionally use linux-only modules
BEGIN {
    if ("$^O" eq "linux") {
        require Linux::Distribution;
    }
}

our $Cmds = {
    dpkg => {
        check => 'dpkg -L "%s"',
        diff => 'apt-get -s install "%s"',
        install => 'apt-get -qq -y install "%s"'
    },
    rpm => {
        check => 'rpm -q "%s"',
        install => 'yum -q -y --noplugins install "%s"'
    },
    emerge => {
        check => 'equery -qC list "%s"',
        diff => 'emerge --color n -pq "%s"',
        install => 'emerge --color n -q "%s"'
    },
    pacman => {
        check => 'pacman -Qe "%s"',
        install => 'pacman -Sq --noconfirm "%s"'
    },
    pkg_info => {
        check => 'pkg_info -Qq "%s>0"'
    }
};

=head1 NAME

Csistck::Test::Pkg - Csistck package check

=head1 METHODS

=head2 pkg($package, [$type])

Test for existing package using forks to system package managers. Package can be
specified as a string, or as a hashref:

  pkg({
      dpkg => 'test-server',
      emerge => 'net-test',
      default => 'test-server'
  });

The package manager will be automatically detected if none is explicitly
specified, and the hashref key matching the package manager decides the package
name to check. If a default key is provided, that package is used by default.

In repair mode, install the package quietly, unless package manager doesn't
handle automating install.

Supported package managers:

=over

=item dpkg

Debian package management utility

=item pacman

Arch linux package management utility

=item More planned..

=back

=cut 

sub pkg { Csistck::Test::Pkg->new($_); };

sub new {
    my ($class, $pkgref, $type) = @_;
    my $self = $class->SUPER::new();
    my $pkg;
    my $diff;

    # Priority: 'type' argument, 'pkg_type' option, detect_pkg_manager.
    if (! $type) {
        $type = option('pkg_type') // detect_pkg_manager();
    }
    die("Bad package manager: manager=<$type>")
      if (! $type);
    die("Package manager not supported: type=<$type>")
      if (! $Cmds->{$type});
    
    $pkg = get_pkg($pkgref, $type);
    $diff = sub { pkg_diff($pkg, $type); }
      if ($Cmds->{$type}->{diff});
    
    $self->{CHECK} = sub { pkg_check($pkg, $type); };
    $self->{REPAIR} = sub { pkg_install($pkg, $type); };
    $self->{DIFF} = $diff;
    $self->{DESC} = "Searching for package $pkg, using $type";

    bless($self, $class);
    return $self;
}

sub pkg_check {
    my ($pkg, $type) = @_;
    my $cmd = sprintf($Cmds->{$type}->{check}, $pkg) or
      die("Package check command missing: type=<$type>");

    debug("Searching for package via command: cmd=<$cmd>");    
    my $ret = system("$cmd 1>/dev/null 2>/dev/null");

    die("Package missing")
      unless($ret == 0);
}

# Package install
sub pkg_install {
    my ($pkg, $type) = @_;
    my $cmd;

    if (defined $Cmds->{$type}->{install}) {
        $cmd = sprintf($Cmds->{$type}->{install}, $pkg);
    }
    else {
        die("Package install command missing: type=<$type>");
    }

    $ENV{DEBIAN_FRONTEND} = "noninteractive";
    debug("Installing package via command: cmd=<$cmd>");
    my $ret = system("$cmd 1>/dev/null 2>/dev/null");

    die("Package installation failed")
      unless ($ret == 0);
}

# Package diff
sub pkg_diff {
    my ($pkg, $type) = @_;
    my $cmd = sprintf($Cmds->{$type}->{diff}, $pkg) or
      die("Package diff command missing: type=<$type>");
    
    $ENV{DEBIAN_FRONTEND} = "noninteractive";
    debug("Showing package differences via command: cmd=<$cmd>");
    my $ret = system("$cmd 2>/dev/null");

    die("Package differences query failed")
      unless ($ret == 0);
}

=head2 detect_pkg_manager()

Detect package manager based on system OS and Linux distribution if
applicable. Return package manager as string. This is not exported, it is 
used for the package test.

=cut

sub detect_pkg_manager {
    given ("$^O") {
        when (/^freebsd$/) { return 'pkg_info'; }
        when (/^netbsd$/) { return 'pkg_info'; }
        when (/^linux$/) { 
            given (Linux::Distribution::distribution_name()) {
                when (/^(?:debian|ubuntu)$/) { return 'dpkg'; }
                when (/^(?:fedora|redhat|centos)$/) { return 'rpm'; }
                when (/^gentoo$/) { return 'emerge'; }
                when (/^arch$/) { return 'pacman'; }
                default { return undef; }
            }
        }
        when (/^darwin$/) { return undef; }
        default { return undef; }
    }
}

=head2 get_pkg($package, $type)

Based on input package, return package name. With OS and distribution detection,
$package can be passed as a string or a hashref.

See pkg() for information on passing hashrefs as package name

=cut

sub get_pkg {
    my $pkg = shift;
    my $type = shift;
    
    debug(sprintf('Searching for package: ref=<%s>, pkg=<%s>', 
      ref $pkg, $pkg));
    
    given (ref $pkg) {
        when ('') {
            return $pkg if ($pkg =~ m/^[A-Za-z0-9\-\_\.\/]+$/);
        }
        when ('HASH') {
            my $pkg_name = $pkg->{$type} // $pkg->{default};
            return $pkg_name if ($pkg_name =~ m/^[A-Za-z0-9\-\_\.\/]+$/);
        }
    }
    
    die('Invalid package');
}

1;
__END__

=head1 OPTIONS

=over

=item pkg_type [string]

Set the default package type

=back

=head1 AUTHOR

Anthony Johnson, C<< <aj@ohess.org> >>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2012 Anthony Johnson

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,



lib/Csistck/Test/File.pm
package Csistck::Test::File;

use 5.010;
use strict;
use warnings;

use base 'Csistck::Test';
use Csistck::Oper qw/debug/;
use Csistck::Util qw/backup_file hash_file/;

our @EXPORT_OK = qw/file/;

use Digest::MD5;
use File::Basename;
use File::Copy;
use FindBin;

sub file {
    my $src = shift;
    my $dest = shift;
    my $args = shift;

    # Get absolute paths for glob, glob files
    my $src_abs = join '/', $FindBin::Bin, $src;
    my @files = glob($src_abs);
    
    # Return array of tests
    return map { Csistck::Test::File->new($_, $dest, $args); } @files;
}

# Helper function to avoid same scope in map and to build test
sub new {
    my ($class, $src, $dest, $args) = @_;
    my $self = $class->SUPER::new();
    # TODO handle args here

    # Test destination for file
    my $src_base = basename($src);
    my $dest_abs = join "/", $dest, $src_base;

    $self->{CHECK} = sub { file_check($src, $dest_abs); };
    $self->{REPAIR} = sub { file_install($src, $dest_abs); };
    $self->{DIFF} = sub { file_diff($src, $dest_abs); };
    $self->{DESC} = "File check on $src";
    
    bless($self, $class);
    return $self;
}

# Check file and file in destination are the same via md5
sub file_check {
    my $src = shift;
    my $dest = shift;

    die("Files do not match: src=<$src> dest=<$dest>")
      unless(file_compare($src, $dest));
}

# Copy file to destination
sub file_install {
    my ($src, $dest) = @_;
    
    backup_file($dest)
      if(-f -e -r $dest);

    debug("Copying file: <src=$src> <dest=$dest>");

    copy($src, $dest) or die("Failed to copy file: $!: src=<$src> dest=<$dest>");
}

# Diff for files
sub file_diff {
    my ($src, $dest) = @_;

    if(-f -e -r $dest) {
        say(Text::Diff::diff($dest, $src));
    }
    else {
        die("Destination file does not exist: dest=<$dest>");
    };
}

# Compare hashes between two files
sub file_compare {
    my @files = @_;

    # Require two files
    return 0 unless (scalar @files == 2);

    # Get hashes and return compare
    my ($hasha, $hashb) = map hash_file($_), @files;

    debug(sprintf "File compare result: <hash=%s> <hash=%s>", $hasha, $hashb);
    
    return ($hasha eq $hashb);
}

1;
__END__

=head1 NAME

Csistck::Test::File - Csistck file check

=head1 DESCRIPTION

=head1 METHODS

=head2 file($glob, $target)

Select files using file glob pattern and copy files into target path. Target
path should be a directory, not a file.

=head1 AUTHOR

Anthony Johnson, E<lt>anthony@ohess.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2011 Anthony Johnson

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.

=cut



lib/Csistck/Oper.pm
package Csistck::Oper;

use 5.010;
use strict;
use warnings;

use base 'Exporter';

our @EXPORT_OK = qw/
    debug error info
    repair
/;

use Carp;
use Getopt::Long;
use Term::ANSIColor;

# Logging levels, default boolean
our $Modes = { 
    debug => 0,
    error => 1,
    info => 0
};

# Boolean options
our $Options = {
    verbose => 0,
    debug => 0,
    quiet => 0,
    repair => 0,
    help => 0,
    color => 1
};

# Set up colored output
$Term::ANSIColor::EACHLINE = "\n";
my $Colors = {
    'debug' => 'black',
    'info' => 'yellow',
    'error' => 'red'
};

# Dynamic setup of reporting functions
for my $level (keys %{$Modes}) {

    no strict 'refs';

    *{"Csistck\::Oper\::$level"} = sub {
        my $func = shift;

        # Maybe this isn't the best way. If func is passed and
        # is code, execute. If func is passed and is a scalar,
        # debug it?!
        if (defined $func and $Modes->{$level}) {
            given (ref $func) {
                when ("CODE") { return &$func; };
                when ("") { return log_message($level, $func) };
            }
        }
        else {
            # Return mode
            return $Modes->{$level};
        }
    };
}

# Repair mode accessor
sub repair {
    return $Options->{repair};
}

# Set up mode via command line options
sub set_mode_by_cli {

    # Map options (as getopt negatable option) to $Options
    # TODO more options
    my %opts = map { +"$_!" => \$Options->{$_} } keys %{$Options};
    my $result = GetOptions(%opts);

    # Set reporting mode based on options
    $Modes->{info} = ($Options->{verbose}) ? 1 : 0;
    $Modes->{debug} = ($Options->{debug}) ? 1 : 0;
    $Modes->{error} = ($Options->{quiet}) ? 0 : 1;
}

# Display usage
sub usage {
    return undef unless ($Options->{help});
    
    print <<EOF;
Usage: $0 [OPTION]...

  Arguments:
    
    --help      Display usage
    --verbose   Verbose output
    --debug     Debug output
    --repair    Run repair operations
    --quiet     Less output
    --nocolor   Turn off colored output

EOF
    return 1;
}

sub log_message {
    my $level = shift;
    my $msg = shift;
    
    my $log_line = sprintf("[%s]\ %s\n", uc($level), $msg);

    if ($Options->{color}) {
        print(colored($log_line, $Colors->{$level}));
    }
    else {
        print($log_line);
    }
}

1;


lib/Csistck/Role.pm
package Csistck::Role;

use strict;
use warnings;
use 5.010;

use Csistck;


sub new {
    my $class = shift;
    my %args = @_;
    my $args = \%args;

    my $self = {};
    $self->{tests} = [];
    bless($self, $class);

    # We won't add keys that overlap with required keys, 
    # drop silently for now. Set up tests after
    $self->defaults();
    foreach my $key (keys ($args)) {
        $self->{$key} = $args->{$key}
          unless ($key =~ /^tests$/);
    }
    $self->tests();

    return $self;
}

# For overriding
sub defaults { return; };
sub tests { return; };

# Add test to object
sub add {
    my $self = shift;
    my @tests = @_;
    push(@{$self->{tests}}, @tests);
}

# Return tests
sub get_tests {
    my $self = shift;
    return $self->{tests};
}

1;
__END__

=head1 NAME

Csistck::Role - Cistck base class for role inheritance

=head1 SYNOPSIS
    
    package Service::Base;

    use base 'Csistck::Role';
    
    sub defaults {
        my $self = shift;
        $self->{some_arg} = 42;
        $self->{config} = '/etc/service.conf';
    }

    sub tests {
        my $self = shift;
        $self->add([
            template(".files/test.tt", "/tmp/test", { service => $self }),
            permission("/tmp/test*", mode => '0777', uid => 100, gid => 100)
        ]);
    }
    
    1;

Create a new instance of the class to add the checks to a host definition:

    host 'example.com' =>
        noop(0),
        Service::Base->new(
            some_arg => 84
        ),
        noop(1);


=head1 DESCRIPTION

This class is the base class for defining custom roles with input arguments.

=head1 METHODS

=head2 defaults()

Override this method on the child class to set argument defaults. Do not 
define methods here, as default arguments are processed before input 
arguments.

=head2 tests()

Override the method on the child class to create tests on the role.

=head1 AUTHOR

Anthony Johnson, E<lt>aj@ohess.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2012 Anthony Johnson

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,




lib/Csistck/Term.pm
package Csistck::Term;

use 5.010;
use strict;
use warnings;

use base 'Exporter';

use Term::ReadKey;
use Csistck::Test;

our @EXPORT_OK = qw/
    prompt
/;

# Interactive terminal prompt. Takes in test as argument, and displays action 
# options
sub prompt {
    my $test = shift;
    my %options;

    # Ask question loop, get input. If choices are not code refs, do not show choice
    say("What would you like to:");
    my @choices = ();
    # Repair
    if ($test->has_repair()) {
        say("  Y : Repair");
        push(@choices, 'Y');
    }
    # Skip
    say("  N : Skip");
    push(@choices, 'n');
    # DIff
    if ($test->has_diff()) {
        say("  D : Diff");
        push(@choices, 'd');
    }
    print("[Y/n/d]? ");
    
    ReadMode 3;
    my $action = ReadKey(0);
    say($action); 
    ReadMode 0;
    
    given ($action) {
        when (/[Yy\n]/) { 
            $test->repair() if ($test->has_repair());
        }
        when (/[Dd]/) { 
            # Show diff, loop through prompt again
            $test->diff() if ($test->has_diff());
            prompt($test);
        }
        default {}
    }
}

1;


lib/Csistck/Util.pm
package Csistck::Util;

use 5.010;
use strict;
use warnings;

use base 'Exporter';
use FindBin;
use File::Copy;

our @EXPORT_OK = qw/
    backup_file
    hash_file
    hash_string
    package_manager
/;

use Csistck::Oper qw/debug info/;
use Csistck::Config qw/option/;

# Backup single file
sub backup_file {
    my $file = shift;

    debug("Backing up file: <file=$file>");
    
    # Get absolute backup path
    my $dest_base = option('backup_path') // join '/', $FindBin::Bin, 'backup';
    die("Backup path does not exist: path=<$dest_base>")
      if (! -e $dest_base);
    die("Backup path is not writable: path=<$dest_base>")
      if (-e $dest_base and ! -w $dest_base);    
    
    # Get file hash, use this is a file name to copy to
    my $hash = hash_file($file);
    my $dest = join '/', $dest_base, $hash;

    copy($file, $dest) or die("Backup failed: $!: file=<$file> dest=<$dest>");

    info("Backup succeeded: file=<$file> dest=<$dest>");
}

# Hash file, return hash or die if error
sub hash_file {
    my $file = shift;

    debug("Hashing file: file=<$file>");

    # Errors to die on
    die("File does not exist: file=<$file>")
      if (! -e $file);
    die("File not readable: file=<$file>")
      if (! -r $file);

    open(my $h, $file) or die("Error opening file: $!: file=<$file>");
        
    my $hash = Digest::MD5->new();
    $hash->addfile($h);
    close($h);

    my $digest = $hash->hexdigest();

    debug(sprintf "File hash successful: file=<%s> hash=<%s>", $file, $digest);

    return $digest;    
}

# Returns md5 hash of input string
sub hash_string {
    my $string = shift;

    my $hash = Digest::MD5->new();
    $hash->add($string);

    return $hash->hexdigest();
}

1;


lib/Csistck.pm
package Csistck;

use 5.010;
use strict;
use warnings;

our $VERSION = '0.09_01';

# We export function in the main namespace
use base 'Exporter';
our @EXPORT = qw(
    host
    role
    check
    option

    file
    noop
    permission
    pkg
    script
    template
);

# Imports for base
use Csistck::Config qw/option/;
use Csistck::Test::NOOP qw/noop/;
use Csistck::Test::File qw/file/;
use Csistck::Test::Permission qw/permission/;
use Csistck::Test::Pkg qw/pkg/;
use Csistck::Test::Script qw/script/;
use Csistck::Test::Template qw/template/;

use Csistck::Role;
use Csistck::Term;

use Sys::Hostname::Long qw//;
use Data::Dumper;
use Scalar::Util qw/blessed reftype/;

# Package wide
my $Hosts = {};
my $Roles = {};


=head1 NAME

Csistck - Perl system consistency check framework

=head1 SYNOPSIS

    use Csistck;
    
    for my $host (qw/a b/) {
        host $host => role('test');
    }

    host 'c' => role('test');
    
    role 'test' => 
        template(".files/test.tt", "/tmp/test", { text => "Some text here" }),
        permission("/tmp/test*", mode => '0777', uid => 100, gid => 100);
    
    check;

The script can then be called directly, using command line arguements below

=head1 DESCRIPTION

Csistck is a small Perl framework for writing scripts to maintain system 
configuration and consistency. The focus of csistck is to stay lightweight,
simple, and flexible.

=head1 EXTENDING ROLES

Roles can be defined using the C<role> keyword syntax, however a more flexible
method is to extend a new object from L<Csistck::Role>:

    use Csistck;
    use base 'Csistck::Role';

    sub defaults {
        my $self = shift;
        $self->{config} = '/etc/example.conf';
    }

    sub tests {
        my $self = shift;
        $self->add(pkg({
            dpkg => 'test-server',
            pkg_info => 'net-test'
        }),
        template('files/example.conf', $self->{config}, { example => $self });
    }

    1;

See L<Csistck::Role> for information on extending roles


=head1 METHODS


=head2 host($host, $checks)

Add tests to host C<$host> test array. Tests are Csistck::Test blessed references, code
references, or arrays of either. To process host tests, use C<check()>.

=cut

sub host {
    my $hostname = shift;

    # Add domain if option is set?
    my $domain_name = Csistck::Config::option('domain_name');
    $hostname = join '.', $hostname, $domain_name
      if (defined $domain_name);

    while (my $require = shift) {
        push(@{$Hosts->{$hostname}}, $require);
    }

    return $Hosts->{$hostname};
}

=head2 role($role, $checks)

Define a weak role. Constructed similar to a host check, however roles are not
called directly, rather they are used to define groups of common tests that can
be used by multiple hosts.

See L<EXTENDING ROLES> above for an object-based style of defining roles, which
allows for passing role configuration.

=cut

sub role {
    my $role = shift;

    # If tests specified, add now
    while (my $require = shift) {
        push(@{$Roles->{$role}}, $require);
    }

    return sub { 
        # Run required role or die
        die ("What's this, \"${role}\"? That role is bupkis.")
          unless (defined $Roles->{$role});
        
        process($Roles->{$role});
    }
}

=head2 check($target)

Runs processing on C<$target>. If C<$target> is C<undef>, then look up the
system's full hostname. If C<$target> is a string, use that string for a
hostname lookup. If C<$target> is a C<Csistck::Test> reference, a coderef, or an
arrayref, then process that object directly. This is useful for writing scripts
where hostname is not important.

=cut

sub check {
    my $target = shift // Sys::Hostname::Long::hostname_long();

    # Process cli arguments for mode/etc, usage
    Csistck::Oper::set_mode_by_cli();
    return if (Csistck::Oper::usage());

    # If target is a string, process as hostname reference. Otherwise, assume a
    # test object was passed
    if (!defined(reftype($target))) {
        die ("What's this, \"${target}\"? That host is bupkis.")
          unless (defined $Hosts->{$target});
        process($Hosts->{$target});
    }
    else {
        process($target);
    }
}

# For recursive testing based on type
sub process {
    my $obj = shift;
    
    # Iterate through array and recursively call process, call code refs, 
    # and run tests

    given (ref $obj) {
        when ("ARRAY") {
            foreach my $subobj (@{$obj}) {
                process($subobj);
            }
        }
        when ("CODE") {
            &{$obj};
        }
        default {
            if (blessed($obj) and $obj->isa('Csistck::Test')) {
                # Check is mandatory, if auto repair is set, repair, otherwise prompt
                if (!$obj->check()) {
                    if (Csistck::Oper::repair()) {
                        $obj->repair()
                    }
                    else {
                        Csistck::Term::prompt($obj);
                    }
                }
            }
            elsif (blessed($obj) and $obj->isa('Csistck::Role')) {
                # Object might be subclass of Csistck::Role
                foreach my $subobj (@{$obj->get_tests()}) {
                    process($subobj);
                }
            }
            else {
                die(sprintf("Unkown object reference: ref=<%s>", ref $obj));
            }
        }
    }
}

1;
__END__

=head1 EXPORTED METHODS

=head2 option($name, $value)

Set option to specified value.

=head3 Available Options

=over 3

=item *

pkg_type [string]

Set package type

=item *

domain_name [string]

Set default domain name to append to hosts

=back

=head2 host($hostname, [@tests]);

Append test or array of tests to host definition.

    host 'hostname' => noop(1), noop(1);
    host 'hostname' => noop(0);

Returns a reference to the host object.

=head2 role($rolename, [@tests]);

Append test or array of tests to role definition.
    
    role 'test' => noop(0);
    host 'hostname' => role('test');

Returns a reference to the role object.


=head2 noop($return)

"No operation" test, used only for testing or placeholders.

    role 'test' => noop(1);

=head2 file($glob, $target)

Copy files matching file glob pattern to target directory. 

    role 'test' => file("lighttpd/app/*.conf", "/etc/lighttpd");

See L<Csistck::Test::File>

=head2 template($template, $target, [%args])

Process file $template as a Template Toolkit template, output to path $target.
Hashref %args is passed to the template processor.

    role 'test' => template("sys/motd", "/etc/motd", { hot => 'dog' });

See L<Csistck::Test::Template>

=head2 permission($glob, %args)

Change permissions on files matching file glob pattern

    role 'test' => permission("/etc/couchdb/*", {
        mode => '0640',
        uid => 130,
        gid => 130
    });

See L<Csistck::Test::Permission>

=head2 script($script, [@arguments])

Call script with specified arguments 

    role 'test' => script("apache2/mod-check", "rewrite");

See L<Csistck::Test::Script>

=head2 pkg($package, [$type])

Check for package using system package manager. The C<package> argument may be
specified as a string, or as a hashref to specify package names for multiple
package managers. The package manager will be automatically detected if no
package manager is specified.

    option 'pkg_type' => 'dpkg';
    role 'test' => 
        pkg("lighttpd", 'dpkg'),
        pkg({
            dpkg => 'snmp-server',
            pkg_info => 'net-snmp'
        });

See L<Csistck::Test::Pkg> for more information

=head1 SCRIPT USAGE

Scripts based on csistck will run in an interactive mode by default. 
The following command line options are recognized in a csistck based script

=over

=item *

B<--[no]repair>

Automate repair mode, do not run in interactive mode.

=item *

B<--[no]verbose>

Toggle verbose reporting of events

=item *

B<--[no]debug>

Toggle debug reporting of events

=item *

B<--[no]quiet>

Toggle event reporting of errors

=back

=head1 AUTHOR

Anthony Johnson, C<< <aj@ohess.org> >>

=head1 COPYRIGHT AND LICENSE

Copyright (c) 2011-2012 Anthony Johnson

Permission is hereby granted, free of charge, to any person obtaining a copy of
this software and associated documentation files (the "Software"), to deal in
the Software without restriction, including without limitation the rights to
use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
the Software, and to permit persons to whom the Software is furnished to do so,
subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

=cut
